chacractl/api/binaries.py

```diff
@@ -71,27 +87,40 @@ def post(self, url, filepath):
             return self.put(file_url, filepath)
         elif exists.status_code == 404:
             logger.info('POSTing file: %s', filepath)
-            with open(filepath, 'rb') as binary:
-                response = requests.post(
+            binary, digest = self.load_file(filepath)
+            response = requests.post(
                     url,
                     files={'file': binary},
                     auth=chacractl.config['credentials'],
                     verify=chacractl.config['ssl_verify'])
         if response.status_code > 201:
             logger.warning("%s -> %s", response.status_code, response.text)
             response.raise_for_status()
+        if not self.verify_upload(url, filename, digest):
+            # Since this is a new file, attempt to delete it
+            logging.error(
+                    'Checksum mismatch: server has wrong checksum for %s! '
+                    'Deleting corrupted file from server...',
+                    filepath)
+            self.delete(file_url)
+            raise SystemExit('Uploaded file checksum mismatch!')
 
     def put(self, url, filepath):
+        filename = os.path.basename(filepath)
         logger.info('resource exists and --force was used, will re-upload')
         logger.info('PUTing file: %s', filepath)
-        with open(filepath, 'rb') as binary:
-            response = requests.put(
+        binary, digest = self.load_file(filepath)
+        response = requests.put(
                 url,
                 files={'file': binary},
                 auth=chacractl.config['credentials'],
                 verify=chacractl.config['ssl_verify'])
         if response.status_code > 201:
             logger.warning("%s -> %s", response.status_code, response.text)
+        if not self.verify_upload(url, filename, digest):
+            # Maybe the old file with a different digest is still there, so
+            # don't delete it
+            raise SystemExit('Uploaded file checksum mismatch!')
```

Comment:
Same thing here, want to log the `filename` in this message instead of "Uploaded file"?